plugin/qbackendconnection.cpp

```diff
@@ -391,44 +381,144 @@ void QBackendConnection::handleMessage(const QByteArray &message)
-            QMetaObject::invokeMethod(this, &QBackendConnection::handlePendingMessages, Qt::QueuedConnection);
+    if (!m_syncResult.isEmpty()) {
+        qCDebug(lcConnection) << "Queueing handling of " << command << " due to syncResult";
+        doDeliver = false;
+    }
+
+    // If the connection is not yet established, then block the message until
+    // callbacks have been installed.
+    if (m_state != ConnectionState::Established && !m_syncCallbacks.size()) {
+        qCDebug(lcConnection) << "Queueing handling of " << command << " as the connection is not yet established, and no callbacks have been installed";
+        doDeliver = false;
+    }
+
+    // Check the command against the state. Don't allow a command to deliver
+    // unless it's in the right state!
+    switch (m_state) {
+    case ConnectionState::WantVersion:
+        if (command != "VERSION") {
+            qCDebug(lcConnection) << "Queueing handling of " << command << " as we are WantVersion";
+            doDeliver = false;
+        }
+        break;
+    case ConnectionState::WantTypes:
+        if (command != "CREATABLE_TYPES") {
+            qCDebug(lcConnection) << "Queueing handling of " << command << " as we are WantTypes";
+            doDeliver = false;
+        }
+        break;
+    case ConnectionState::WantEngine:
+        // Need a QML engine pointer. Queue everything except further ROOT.
+        qCDebug(lcConnection) << "Queueing handling of " << command << " as we WantEngine";
+        doDeliver = false;
+        break;
+    case ConnectionState::WantRoot:
+        if (command != "ROOT") {
+            qCDebug(lcConnection) << "Queueing handling of " << command << " as we are WantRoot";
+            doDeliver = false;
         }
+        break;
+    case ConnectionState::Established:
+        break; // no queueing based on state.
+    }
+
+    if (!doDeliver) {
         m_pendingMessages.append(cmd);
         return;
     }
 
-    if (m_syncCallback) {
-        // Reset callback, set result, and handle immediately, bypassing the queue
+    // Pass the message to all sync callbacks to see if any of them want it.
+    // If so, handle it immediately, bypassing the queue.
+    bool wasHandled = false;
+    for (int i = 0; i < m_syncCallbacks.size(); i++) {
+        // Remove callback, set result, and handle immediately, bypassing the queue
         // even if out of order; see waitForMessage for details.
-        m_syncCallback = nullptr;
-        m_syncResult = cmd;
-    } else if (!m_pendingMessages.isEmpty()) {
-        // Other messages are pending, so queue this one as well
-        qCDebug(lcConnection) << "Queuing handling of" << command << "into non-empty message queue";
-        // In this case, there's no need to trigger the handler later; something already has
-        m_pendingMessages.append(cmd);
-        return;
+        if (m_syncCallbacks.at(i)(cmd)) {
+            m_syncCallbacks.removeAt(i);
+            m_syncResult = cmd;
```

Comment:
This can return the wrong result.

`waitForMessage` calls must always be handled in LIFO order. Logically, that makes sense: we're actually blocking the thread, so any nested `waitForMessage` is blocking that blocking call and it needs to be unblocked first.

`m_syncResult` works with that in mind: it's the result for the `waitForMessage` that directly led to this message being handled. When `waitForReadyRead` returns, it will unblock/return if there is a result. It saves and restores `m_syncResult` so that nested calls will work correctly.

I think this block needs to revert to what it was, and if that is causing problems anywhere the bug isn't here.